### PR TITLE
Create first version of logMultiple for Program and Console loggers

### PR DESCRIPTION
**Pull request by anagperal**
### :pushpin: References

-   **Issue:** Closes #8694u3jqc [Create first version log multiple messages](https://app.clickup.com/t/8694u3jqc)

### :memo: Implementation
- Create the first version to log multiple messages at once in ProgramLogger and ConsoleLogger

### CODE REVIEW

File: src/domain/usecases/LogMultipleMessageUseCase.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import { Log } from "../../domain/entities/Log";
+import { LoggerRepository } from "../../domain/repositories/LoggerRepository";
+
+export class LogMultipleMessageUseCase {
+    constructor(private loggerRepository: LoggerRepository) {}
+
+    public execute(logs: Log[], options?: { isDebug?: boolean }): FutureData<void> {
+        if (options?.isDebug) {
+            // eslint-disable-next-line no-console
+            console.log(logs);
```
Comment (9sneha-n) [inline src/domain/usecases/LogMultipleMessageUseCase.ts]: if you do console.debug, you wont get a ts error/warning. low importance.

File: src/index.ts [hunk 2]
```diff
@@ -27,6 +28,7 @@ export interface Logger<T> {
     success(content: T): Promise<void>;
     warn(content: T): Promise<void>;
     error(content: T): Promise<void>;
+    logMultiple(content: MultipleLogContent): Promise<void>;
```
Comment (9sneha-n) [inline src/index.ts]: we can think about the best name for this. Is batchLog() better than logMultiple? Low importance

**Review by 9sneha-n — APPROVED** [review]
Looks good @anagperal !

File: src/data/repositories/ConsoleLoggerRepository.ts [hunk 3]
```diff
@@ -4,6 +4,10 @@ import { DefaultLog } from "../../domain/entities/Log";
 import { LoggerRepository } from "../../domain/repositories/LoggerRepository";
 
 export class ConsoleLoggerRepository implements LoggerRepository {
+    batchLog(logs: DefaultLog[]): FutureData<void> {
+        logs.map(log => this.log(log));
+        return Future.success(undefined);
+    }
```
Comment (tokland) [inline src/data/repositories/ConsoleLoggerRepository.ts:10]: The method is async (it returns a future) so, even if the action to perform (writing to console) is not async itself, it has to be wrapped somehow within a Future. Otherwise, we are not honoring a basic contract of Future: it's only executed on `.run`. A proposal:

```
        const logs$ = logs.map(log => this.log(log));
        return Future.sequential(logs$).map(() => undefined);
```

Note that `log` bellow also needs to be refactored. Something like this:

```
        return Future.success<Error, void>(undefined).tap(() => {
            const date = new Date().toISOString();
            process.stderr.write(`[${messageType.toUpperCase()}] [${date}] ${message}\n`);
        });
```

(get `Future.tap` from some other EST projects)

File: src/data/repositories/ConsoleLoggerRepository.ts [hunk 4]
```diff
@@ -4,6 +4,10 @@ import { DefaultLog } from "../../domain/entities/Log";
...
     log(log: DefaultLog): FutureData<void> {
```
Comment (tokland) [inline src/data/repositories/ConsoleLoggerRepository.ts:12]: prettier does not enforce it, but multi-line methods are typically separated by a blank link to improve readability. This does not apply to interfaces (one-liners).

File: src/data/repositories/ProgramLoggerD2Repository.ts [hunk 5]
```diff
@@ -25,20 +27,45 @@ export class ProgramLoggerD2Repository implements LoggerRepository {
         this.messageId = dataElements.messageId;
         this.messageTypeId = dataElements.messageTypeId;
         this.organisationUnitId = organisationUnitId;
+        this.programStage = undefined;
+    }
+
+    async init(): Promise<void> {
+        this.programStage = await this.getProgramStage().toPromise();
     }
```
Comment (tokland) [inline src/data/repositories/ProgramLoggerD2Repository.ts:57]: `this.programStage = undefined;` raises a flag: imperative pattern. I see the problem, you have some async initialization which cannot be in the constructor. The typical solution for this would to be:

- Add "programStage" as argument to the constructor
- static async init(): Promise<void> { ... return new ProgramLoggerD2Repository(...) }

I think this will work here. Sometimes this pattern does not abide by the caller. Then we'd be forced to make `this.programStage` a property (cached) which returns a FutureData<D2ProgramStage> -> that requires to refactor methods using it., so let's see if option (1) works for you.

File: src/data/repositories/ProgramLoggerD2Repository.ts [hunk 6]
```diff
@@ -25,20 +27,45 @@ export class ProgramLoggerD2Repository implements LoggerRepository {
...
+        }
+
+        const d2EventProgram = this.mapLogToD2EventProgam({
+            log,
+            programId: this.programId,
+            organisationUnitId: this.organisationUnitId,
+            messageId: this.messageId,
+            messageTypeId: this.messageTypeId,
+            programStage: this.programStage,
+        });
```
Comment (tokland) [inline src/data/repositories/ProgramLoggerD2Repository.ts:49]: [subjective] You can pass `...this` as first item to make it shorter.

File: src/data/repositories/ProgramLoggerD2Repository.ts [hunk 7]
```diff
@@ -25,20 +27,45 @@ export class ProgramLoggerD2Repository implements LoggerRepository {
...
+        return this.postApiTracker([d2EventProgram]);
+    }
+
+    batchLog(logs: DefaultLog[]): FutureData<void> {
```
Comment (tokland) [inline src/data/repositories/ProgramLoggerD2Repository.ts:71]: I agree with Sneha, `logMultiple` sounded weird.

File: src/data/repositories/ProgramLoggerD2Repository.ts [hunk 8]
```diff
@@ -25,20 +27,45 @@ export class ProgramLoggerD2Repository implements LoggerRepository {
...
+    }
+
+    async init(): Promise<void> {
+        this.programStage = await this.getProgramStage().toPromise();
     }
 
     log(log: DefaultLog): FutureData<void> {
-        return this.getProgramStage().flatMap(programStage => {
-            const d2EventProgram = this.mapLogToD2EventProgam({
+        if (!this.programStage) {
```
Comment (tokland) [inline src/data/repositories/ProgramLoggerD2Repository.ts:60]: And once refactored, all this this.programStage checks go away (the type will have not be a `Maybe`)

File: src/data/repositories/TrackerProgramLoggerD2Repository.ts [hunk 9]
```diff
@@ -25,6 +30,11 @@ export class TrackerProgramLoggerD2Repository implements LoggerRepository {
         this.messageTypeId = messageTypeId;
     }
 
+    // TODO: implement batchLog method
+    batchLog(_logs: Log[]): FutureData<void> {
+        return Future.success(undefined);
+    }
+
```
Comment (tokland) [inline src/data/repositories/TrackerProgramLoggerD2Repository.ts:37]: Just as a note, a simple basic implementation of a batchLog is Future.sequential of a map of the single log.

File: src/data/repositories/TrackerProgramLoggerD2Repository.ts [hunk 10]
```diff
@@ -25,6 +30,11 @@ export class TrackerProgramLoggerD2Repository implements LoggerRepository {
...
     log(log: TrackerProgramLog): FutureData<void> {
```
Comment (tokland) [inline src/data/repositories/TrackerProgramLoggerD2Repository.ts:38]: It does not belong in this PR, but that signature does not look right. We are implementing a `LoggerRepository`, which has the method defined as `log(log: Log): FutureData<void>`.  But then the type here is `TrackerProgramLog`. And if it's changed to `Log`, type errors raise. So something is amiss. (IMO, this should have been caught by TS, there are some discussions in their github issues)

File: src/domain/entities/BatchLogContent.ts [hunk 11]
```diff
@@ -0,0 +1,3 @@
+import { MessageType } from "./Log";
+
+export type BatchLogContent = { content: string; messageType: MessageType }[];
```
Comment (tokland) [inline src/domain/entities/BatchLogContent.ts]: The canonical advice is: use `SomeType[]` only if something has no spaces. Workarounds: Array<...>, or better, create a separate type for the inner type, and then use it `SomeType[]`

File: src/loggers/ConsoleLogger.ts [hunk 12]
```diff
@@ -32,6 +34,17 @@ export class ConsoleLogger implements Logger<string> {
         return this.log(content, "Error");
     }
 
+    batchLog(content: BatchLogContent): Promise<void> {
+        if (this.loggerRepository) {
```
Comment (tokland) [inline src/loggers/ConsoleLogger.ts]: Why the `if`? The type is `LoggerRepository`.

File: src/loggers/ConsoleLogger.ts [hunk 13]
```diff
@@ -32,6 +34,17 @@ export class ConsoleLogger implements Logger<string> {
...
+            const logs = content.map(content =>
+                this.mapContentToLog(content.content, content.messageType)
+            );
+            return new BatchLogMessageUseCase(this.loggerRepository).execute(logs).toPromise();
```
Comment (tokland) [inline src/loggers/ConsoleLogger.ts]: Typically we use all Promises or all Futures. I see that Logger uses Promises. Let's not change it, but just as a note.

File: src/loggers/ProgramLogger.ts [hunk 14]
```diff
@@ -44,14 +47,31 @@ export class ProgramLogger implements Logger<string> {
...
+            return new BatchLogMessageUseCase(this.loggerRepository)
+                .execute(logs, options)
                 .toPromise();
         } else {
             throw new Error(`Logger not initialized properly. Please check configuration.`);
         }
     }
+
+    private log(content: string, messageType: MessageType): Promise<void> {
+        if (this.loggerRepository) {
```
Comment (tokland) [inline src/loggers/ProgramLogger.ts]: Same

File: src/loggers/ProgramLogger.ts [hunk 15]
```diff
@@ -44,14 +47,31 @@ export class ProgramLogger implements Logger<string> {
...
+            const log = this.mapContentToLog(content, messageType);
+            return new LogMessageUseCase(this.loggerRepository).execute(log, options).toPromise();
+        } else {
+            throw new Error(`Logger not initialized properly. Please check configuration.`);
+        }
+    }
+
+    private mapContentToLog(content: string, messageType: MessageType): DefaultLog {
+        return { message: content, messageType: messageType };
+    }
```
Comment (tokland) [inline src/loggers/ProgramLogger.ts:61]: ConsoleLogger and ProgramLogger have some duplicated code. Unless this is accidental duplication, they should be DRYed using some common module.

**Review by tokland — COMMENTED** [review]
[code-only review]

Code looking good overall! Some implementation details:

File: src/data/repositories/TrackerProgramLoggerD2Repository.ts (hunk 9 above)
Comment (anagperal) [reply src/data/repositories/TrackerProgramLoggerD2Repository.ts:37]: That sounds good! When there is time to implement also the batchLog for TrackerProgramLogger, we can implement this idea in all the loggers

File: src/data/repositories/TrackerProgramLoggerD2Repository.ts (hunk 10 above)
Comment (anagperal) [reply src/data/repositories/TrackerProgramLoggerD2Repository.ts:38]: Maybe can we check this in the next Technical decisions meeting?

**Review by tokland — APPROVED** [review]
:+1:  to the new fixes. I guess we'll tackle the thing with log(log: TrackerProgramLog) in the future. An implementation should have the same signature that the interface.

**Comment by anagperal** [comment]
> 👍 to the new fixes. I guess we'll tackle the thing with log(log: TrackerProgramLog) in the future. An implementation should have the same signature that the interface.

Thanks Arnau! totally agree that we have to solve this issue

**Review by bhavananarayanan — COMMENTED** [review]
Thanks, approved
